### PR TITLE
Started extending IRGen to be able to run a graph_op.

### PR DESCRIPTION
**Pull request by mhong**
The new code path is protected by flag "tf-dynamic-compilation".

This PR calls some new compiler runtime entry points to run the graph_op insts
via eager C APIs, and it only supports computing a fixed float scalar const tensor.

These entry points will be generalized in later PRs to run arbitrary TF tensor computation.

### CODE REVIEW

**Comment by mhong** [comment]
This PR will be extended (and tested) once we submit the new TF experimental C API TFE_RunConstOp() that it depends on.

File: lib/IRGen/IRGenSIL.cpp [hunk 1]
```diff
@@ -23,33 +23,35 @@
 #include "llvm/IR/IntrinsicInst.h"
 #include "llvm/IR/InlineAsm.h"
 #include "llvm/IR/Intrinsics.h"
-#include "llvm/ADT/MapVector.h"
```
Comment (mhong) [inline lib/IRGen/IRGenSIL.cpp]: I only added the following to this section, and then ran clang-format, which reordered the entries.

```
// SWIFT_ENABLE_TENSORFLOW
#include "llvm/IR/TypeBuilder.h"
```

File: lib/IRGen/IRGenSIL.cpp [hunk 2]
```diff
@@ -1861,6 +1866,44 @@ void IRGenSILFunction::visitSILBasicBlock(SILBasicBlock *BB) {
...
+  // The overall workflow is:
+  // 1. Prepare the input tensor handles and attributes
+  // 2. Run the graph_op
+  // 3. Set the output tensor handles via setLoweredExplosion()
+
+  // The true return type is TFE_Context*, which is an opaque pointer, so it
+  // maps to void* in the Swift-C calling convention.
+  auto getContextFn = llvmModule.getOrInsertFunction(
+      "_swift_tfc_GetGlobalEagerContext",
+      llvm::TypeBuilder<void *(), false>::get(llvmContext));
```
Comment (bgogul) [inline lib/IRGen/IRGenSIL.cpp]: nit: `, /*cross_compilable=*/false>`

**Review by bgogul — COMMENTED** [review]
Nice!

File: lib/SILOptimizer/Mandatory/TFPartition.cpp [hunk 3]
```diff
@@ -4555,7 +4569,9 @@ bool TFPartition::partitionFunction(
   // isolation.
   LLVM_DEBUG(llvm::dbgs() << "Processing SIL function " << hostFn->getName()
                           << " in TFPartition::partitionFunction().\n");
-  if (!tfc.shouldBePartitioned(hostFn, /*forceTFFunctions=*/true))
+
+  if (!tfc.shouldBePartitioned(hostFn, /*forceTFFunctions=*/true) ||
```
[the remark this answers is not in the document]
Comment (mhong) [reply lib/SILOptimizer/Mandatory/TFPartition.cpp]: When we do early exit here, graph lowering should be skipped, so we are flowing all graph_op insts to IRGen. What behavior difference are you referring to?

File: lib/IRGen/IRGenSIL.cpp (hunk 2 above)
Comment (mhong) [reply lib/IRGen/IRGenSIL.cpp]: Will add.

File: lib/SILOptimizer/Mandatory/TFPartition.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (bgogul) [reply lib/SILOptimizer/Mandatory/TFPartition.cpp]: Ah, graph lowering never gets invoked if we return here. Sounds good.

**Review by bgogul — APPROVED** [review]

**Comment by mhong** [comment]
@swift-ci please test tensorflow

**Comment by mhong** [comment]
@swift-ci please clean test tensorflow

**Comment by mhong** [comment]
Tests passed (with some flaky test in GPU run).

I'm merging to make forward progress. Please feel free to provide more feedback.

File: stdlib/public/TensorFlow/CompilerRuntime.swift [hunk 4]
```diff
@@ -1145,3 +1145,24 @@ public func _TFCCreateCTensorHandle<T>(_ value : T,
...
+@_silgen_name("_swift_tfc_RunEagerConstTest")
+public func _RunEagerConstTest(_ ctx: CTFEContext) -> TensorHandle<Float> {
+  debugLog("Calling _RunEagerConstTest()")
+  let cHandle: CTensorHandle! = TFE_RunConstOp(ctx)
+  return TensorHandle<Float>(owning: cHandle)
+}
+
+@inlinable
+@_silgen_name("_swift_tfc_GetGlobalEagerContext")
+public func _GetGlobalEagerContext() -> CTFEContext {
```
Comment (rxwei) [inline stdlib/public/TensorFlow/CompilerRuntime.swift]: I don’t think these have to be public. @usableFromInline should do.

**Pull request merged**